pyOCD/rtos/freertos.py

```diff
@@ -0,0 +1,499 @@
+        listsToRead.append((self._symbols['xPendingReadyList'], FreeRTOSThread.READY))
+        if self._symbols.has_key('xSuspendedTaskList'):
+            listsToRead.append((self._symbols['xSuspendedTaskList'], FreeRTOSThread.SUSPENDED))
+        if self._symbols.has_key('xTasksWaitingTermination'):
+            listsToRead.append((self._symbols['xTasksWaitingTermination'], FreeRTOSThread.DELETED))
+
+        for listPtr, state in listsToRead:
+            for threadBase in TargetList(self._target_context, listPtr):
+                try:
+                    # Don't try adding more threads than the number of threads that FreeRTOS says there are.
+                    if len(newThreads) >= threadCount:
+                        break
+
+                    # Reuse existing thread objects.
+                    if threadBase in self._threads:
+                        t = self._threads[threadBase]
+                    else:
+                        t = FreeRTOSThread(self._target_context, self, threadBase)
+
+                    # Set thread state.
+                    if threadBase == currentThread:
+                        t.state = FreeRTOSThread.RUNNING
+                    else:
+                        t.state = state
+
+                    log.debug("Thread 0x%08x (%s)", threadBase, t.name)
+                    newThreads[t.unique_id] = t
+                except DAPAccess.TransferError:
+                    log.debug("TransferError while examining thread 0x%08x", threadBase)
+
+        if len(newThreads) != threadCount:
+            log.warning("FreeRTOS: thread count mismatch")
+
+        # Create fake handler mode thread.
+        if self.get_ipsr() > 0:
+            log.debug("FreeRTOS: creating handler mode thread")
+            t = HandlerModeThread(self._target_context, self)
+            newThreads[t.unique_id] = t
+
+        self._threads = newThreads
+
+    def get_threads(self):
+        if not self.is_enabled:
+            return []
+        self.update_threads()
+        return self._threads.values()
+
+    def get_thread(self, threadId):
+        if not self.is_enabled:
+            return None
+        self.update_threads()
+        return self._threads.get(threadId, None)
+
+    @property
+    def is_enabled(self):
+        return self._symbols is not None and self.get_is_running()
+
+    @property
+    def current_thread(self):
+        if not self.is_enabled:
+            return None
+        self.update_threads()
+        id = self.get_current_thread_id()
+        try:
+            return self._threads[id]
+        except KeyError:
+            return None
+
+    def is_valid_thread_id(self, threadId):
+        if not self.is_enabled:
+            return False
+        self.update_threads()
+        return threadId in self._threads
+
+    def get_current_thread_id(self):
+        if not self.is_enabled:
+            return None
+        if self.get_ipsr() > 0:
+            return 2
```

Comment:
It just needs an arbitrary, fixed thread ID so gdb will recognize it as the same thread. The default Thread mode (before threading is enabled) thread ID is 1.